File: backend/app/services/survey_logic/expression_engine.py
"""ExpressionScript evaluator for server-side survey logic."""

from __future__ import annotations

import math
import random
import re
from dataclasses import dataclass
from typing import Any


class ExpressionError(Exception):
    pass


@dataclass
class EvaluationContext:
    participant_responses: dict[str, Any]
    panel_metadata: dict[str, Any]
    system_variables: dict[str, Any]


_RESERVED = frozenset(
    {"if", "sum", "count", "rand", "array_filter", "and", "or", "not", "true", "false"}
)


def _to_bool(value: Any) -> bool:
    if isinstance(value, bool):
        return value
    if value is None or value == "":
        return False
    if isinstance(value, (list, tuple)):
        return len(value) > 0
    if isinstance(value, (int, float)):
        return value != 0
    return True


def _to_number(value: Any) -> float:
    if isinstance(value, bool):
        return 1.0 if value else 0.0
    if isinstance(value, (int, float)):
        return float(value)
    try:
        return float(value)
    except (TypeError, ValueError):
        return 0.0


def _resolve(name: str, ctx: EvaluationContext) -> Any:
    if name in ctx.participant_responses:
        return ctx.participant_responses[name]
    if name in ctx.panel_metadata:
        return ctx.panel_metadata[name]
    if name in ctx.system_variables:
        return ctx.system_variables[name]
    return None


def _tokenize(expr: str) -> list[tuple[str, str]]:
    tokens: list[tuple[str, str]] = []
    i = 0
    s = expr.strip()
    while i < len(s):
        if s[i].isspace():
            i += 1
            continue
        matched = False
        for kw in ("==", "!=", ">=", "<=", "&&", "||"):
            if s.startswith(kw, i):
                tokens.append(("op", kw))
                i += len(kw)
                matched = True
                break
        if matched:
            continue
        if s[i] in "+-*/<>=!(),":
            tokens.append(("op", s[i]))
            i += 1
            continue
        if s[i] in "\"'":
            q = s[i]
            j = i + 1
            buf: list[str] = []
            while j < len(s) and s[j] != q:
                buf.append(s[j])
                j += 1
            if j >= len(s):
                raise ExpressionError("Unterminated string")
            tokens.append(("str", "".join(buf)))
            i = j + 1
            continue
        if s[i].isdigit() or (s[i] == "." and i + 1 < len(s) and s[i + 1].isdigit()):
            j = i
            while j < len(s) and (s[j].isdigit() or s[j] == "."):
                j += 1
            tokens.append(("num", s[i:j]))
            i = j
            continue
        if s[i].isalpha() or s[i] == "_":
            j = i
            while j < len(s) and (s[j].isalnum() or s[j] == "_"):
                j += 1
            tokens.append(("id", s[i:j]))
            i = j
            continue
        raise ExpressionError(f"Unexpected character '{s[i]}' at {i}")
    tokens.append(("eof", ""))
    return tokens


class _Parser:
    def __init__(self, tokens: list[tuple[str, str]], ctx: EvaluationContext):
        self.tokens = tokens
        self.pos = 0
        self._ctx = ctx

    def _peek(self) -> tuple[str, str]:
        if self.pos >= len(self.tokens):
            return ("eof", "")
        return self.tokens[self.pos]

    def _eat(self) -> tuple[str, str]:
        t = self.tokens[self.pos]
        self.pos += 1
        return t

    def _at_eof(self) -> bool:
        return self._peek()[0] == "eof"

    def parse(self) -> Any:
        return self._or()

    def _or(self) -> Any:
        left = self._and()
        while not self._at_eof() and self._peek()[1] in ("or", "||"):
            self._eat()
            left = _to_bool(left) or _to_bool(self._and())
        return left

    def _and(self) -> Any:
        left = self._not()
        while not self._at_eof() and self._peek()[1] in ("and", "&&"):
            self._eat()
            left = _to_bool(left) and _to_bool(self._not())
        return left

    def _not(self) -> Any:
        if not self._at_eof() and self._peek()[1] in ("not", "!"):
            self._eat()
            return not _to_bool(self._not())
        return self._compare()

    def _compare(self) -> Any:
        left = self._add()
        ops = {"==", "!=", ">", "<", ">=", "<="}
        while not self._at_eof() and self._peek()[1] in ops:
            op = self._eat()[1]
            right = self._add()
            if op == "==":
                left = str(left) == str(right)
            elif op == "!=":
                left = str(left) != str(right)
            elif op == ">":
                left = _to_number(left) > _to_number(right)
            elif op == "<":
                left = _to_number(left) < _to_number(right)
            elif op == ">=":
                left = _to_number(left) >= _to_number(right)
            elif op == "<=":
                left = _to_number(left) <= _to_number(right)
        return left

    def _add(self) -> Any:
        left = self._mul()
        while not self._at_eof() and self._peek()[1] in "+-":
            op = self._eat()[1]
            right = self._mul()
            left = _to_number(left) + _to_number(right) if op == "+" else _to_number(left) - _to_number(right)
        return left

    def _mul(self) -> Any:
        left = self._unary()
        while not self._at_eof() and self._peek()[1] in "*/":
            op = self._eat()[1]
            right = self._unary()
            left = _to_number(left) * _to_number(right) if op == "*" else _to_number(left) / _to_number(right)
        return left

    def _unary(self) -> Any:
        if not self._at_eof() and self._peek()[1] == "-":
            self._eat()
            return -_to_number(self._unary())
        return self._primary()

    def _primary(self) -> Any:
        kind, val = self._peek()
        if kind == "num":
            self._eat()
            return float(val) if "." in val else int(val)
        if kind == "str":
            self._eat()
            return val
        if kind == "id":
            self._eat()
            if val == "true":
                return True
            if val == "false":
                return False
            if self._peek()[1] == "(":
                self._eat()
                args: list[Any] = []
                if self._peek()[1] != ")":
                    args.append(self._or())
                    while self._peek()[1] == ",":
                        self._eat()
                        args.append(self._or())
                if self._peek()[1] != ")":
                    raise ExpressionError("Expected )")
                self._eat()
                return _call(val, args)
            return _resolve(val, self._ctx)
        if val == "(":
            self._eat()
            inner = self._or()
            if self._peek()[1] != ")":
                raise ExpressionError("Expected )")
            self._eat()
            return inner
        raise ExpressionError(f"Unexpected token {val}")

    @property
    def ctx(self) -> EvaluationContext:
        return self._ctx


def _call(name: str, args: list[Any]) -> Any:
    if name == "if":
        return args[1] if _to_bool(args[0]) else args[2]
    if name == "sum":
        return sum(_to_number(a) for a in args)
    if name == "count":
        return sum(1 for a in args if _to_bool(a))
    if name == "rand":
        return random.randint(int(_to_number(args[0])), int(_to_number(args[1])))
    if name == "array_filter":
        arr = args[0] if isinstance(args[0], list) else []
        pred = args[1] if len(args) > 1 else None
        return [x for x in arr if str(x) == str(pred)]
    raise ExpressionError(f"Unknown function {name}")


def evaluate_expression(expression: str, context: EvaluationContext | dict[str, Any]) -> Any:
    trimmed = (expression or "").strip()
    if not trimmed:
        return True
    if isinstance(context, dict) and "participant_responses" not in context:
        ctx = EvaluationContext(
            participant_responses=context,
            panel_metadata={},
            system_variables={},
        )
    else:
        ctx = context  # type: ignore[assignment]
    parser = _Parser(_tokenize(trimmed), ctx)
    return parser.parse()


def interpolate_text(template: str, context: EvaluationContext) -> str:
    def repl(match: re.Match[str]) -> str:
        try:
            result = evaluate_expression(match.group(1).strip(), context)
            return "" if result is None else str(result)
        except ExpressionError:
            return match.group(0)

    return re.sub(r"\{([^{}]+)\}", repl, template)


def is_relevant(expression: str | None, context: EvaluationContext) -> bool:
    if not (expression or "").strip():
        return True
    try:
        return _to_bool(evaluate_expression(expression, context))
    except ExpressionError:
        return False

File: frontend/src/lib/surveyLogic/expressionEngine.ts
/**
 * ExpressionScript evaluator — tokenizer/parser for survey logic and micro-tailoring.
 * Supports: == != && || and or not > < >= <= + - * /
 * Functions: if(), sum(), count(), rand(), array_filter()
 */

import type { EvaluationContext } from './types'

export class ExpressionError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ExpressionError'
  }
}

type TokenType =
  | 'number'
  | 'string'
  | 'identifier'
  | 'operator'
  | 'paren'
  | 'comma'
  | 'eof'

interface Token {
  type: TokenType
  value: string
  pos: number
}

const RESERVED_WORDS = new Set([
  'if',
  'sum',
  'count',
  'rand',
  'array_filter',
  'and',
  'or',
  'not',
  'true',
  'false',
])

function tokenize(input: string): Token[] {
  const tokens: Token[] = []
  let i = 0
  const s = input.trim()
  while (i < s.length) {
    const ch = s[i]
    if (/\s/.test(ch)) {
      i++
      continue
    }
    if (ch === '"' || ch === "'") {
      const quote = ch
      let j = i + 1
      let value = ''
      while (j < s.length && s[j] !== quote) {
        if (s[j] === '\\' && j + 1 < s.length) {
          value += s[j + 1]
          j += 2
        } else {
          value += s[j]
          j++
        }
      }
      if (j >= s.length) throw new ExpressionError(`Unterminated string at ${i}`)
      tokens.push({ type: 'string', value, pos: i })
      i = j + 1
      continue
    }
    if (/[0-9.]/.test(ch)) {
      let j = i
      while (j < s.length && /[0-9.]/.test(s[j])) j++
      tokens.push({ type: 'number', value: s.slice(i, j), pos: i })
      i = j
      continue
    }
    if (/[a-zA-Z_]/.test(ch)) {
      let j = i
      while (j < s.length && /[a-zA-Z0-9_]/.test(s[j])) j++
      tokens.push({ type: 'identifier', value: s.slice(i, j), pos: i })
      i = j
      continue
    }
    const two = s.slice(i, i + 2)
    const three = s.slice(i, i + 3)
    if (['==', '!=', '>=', '<=', '&&', '||'].includes(two)) {
      tokens.push({ type: 'operator', value: two, pos: i })
      i += 2
      continue
    }
    if (['and', 'or', 'not'].includes(three)) {
      tokens.push({ type: 'identifier', value: three, pos: i })
      i += 3
      continue
    }
    if ('+-*/<>=!'.includes(ch)) {
      tokens.push({ type: 'operator', value: ch, pos: i })
      i++
      continue
    }
    if (ch === '(' || ch === ')') {
      tokens.push({ type: 'paren', value: ch, pos: i })
      i++
      continue
    }
    if (ch === ',') {
      tokens.push({ type: 'comma', value: ch, pos: i })
      i++
      continue
    }
    throw new ExpressionError(`Unexpected character '${ch}' at position ${i}`)
  }
  tokens.push({ type: 'eof', value: '', pos: s.length })
  return tokens
}

type AstNode =
  | { kind: 'literal'; value: string | number | boolean }
  | { kind: 'identifier'; name: string }
  | { kind: 'unary'; op: string; arg: AstNode }
  | { kind: 'binary'; op: string; left: AstNode; right: AstNode }
  | { kind: 'call'; name: string; args: AstNode[] }

class Parser {
  private tokens: Token[]
  private pos = 0

  constructor(tokens: Token[]) {
    this.tokens = tokens
  }

  private peek(): Token {
    return this.tokens[this.pos] ?? { type: 'eof', value: '', pos: 0 }
  }

  private consume(): Token {
    return this.tokens[this.pos++] ?? { type: 'eof', value: '', pos: 0 }
  }

  parse(): AstNode {
    const node = this.parseOr()
    if (this.peek().type !== 'eof') {
      throw new ExpressionError(`Unexpected token '${this.peek().value}'`)
    }
    return node
  }

  private parseOr(): AstNode {
    let left = this.parseAnd()
    while (
      (this.peek().type === 'identifier' && this.peek().value === 'or') ||
      (this.peek().type === 'operator' && this.peek().value === '||')
    ) {
      this.consume()
      left = { kind: 'binary', op: '||', left, right: this.parseAnd() }
    }
    return left
  }

  private parseAnd(): AstNode {
    let left = this.parseNot()
    while (
      (this.peek().type === 'identifier' && this.peek().value === 'and') ||
      (this.peek().type === 'operator' && this.peek().value === '&&')
    ) {
      this.consume()
      left = { kind: 'binary', op: '&&', left, right: this.parseNot() }
    }
    return left
  }

  private parseNot(): AstNode {
    if (this.peek().type === 'identifier' && this.peek().value === 'not') {
      this.consume()
      return { kind: 'unary', op: 'not', arg: this.parseNot() }
    }
    if (this.peek().type === 'operator' && this.peek().value === '!') {
      this.consume()
      return { kind: 'unary', op: 'not', arg: this.parseNot() }
    }
    return this.parseComparison()
  }

  private parseComparison(): AstNode {
    let left = this.parseAdd()
    const ops = ['==', '!=', '>=', '<=', '>', '<']
    while (this.peek().type === 'operator' && ops.includes(this.peek().value)) {
      const op = this.consume().value
      left = { kind: 'binary', op, left, right: this.parseAdd() }
    }
    return left
  }

  private parseAdd(): AstNode {
    let left = this.parseMul()
    while (this.peek().type === 'operator' && '+-'.includes(this.peek().value)) {
      const op = this.consume().value
      left = { kind: 'binary', op, left, right: this.parseMul() }
    }
    return left
  }

  private parseMul(): AstNode {
    let left = this.parseUnary()
    while (this.peek().type === 'operator' && '*/'.includes(this.peek().value)) {
      const op = this.consume().value
      left = { kind: 'binary', op, left, right: this.parseUnary() }
    }
    return left
  }

  private parseUnary(): AstNode {
    if (this.peek().type === 'operator' && this.peek().value === '-') {
      this.consume()
      return { kind: 'unary', op: '-', arg: this.parseUnary() }
    }
    return this.parsePrimary()
  }

  private parsePrimary(): AstNode {
    const t = this.peek()
    if (t.type === 'number') {
      this.consume()
      return { kind: 'literal', value: Number(t.value) }
    }
    if (t.type === 'string') {
      this.consume()
      return { kind: 'literal', value: t.value }
    }
    if (t.type === 'identifier') {
      const name = this.consume().value
      if (name === 'true') return { kind: 'literal', value: true }
      if (name === 'false') return { kind: 'literal', value: false }
      if (RESERVED_WORDS.has(name) && this.peek().type !== 'paren') {
        throw new ExpressionError(`Unexpected keyword ${name}`)
      }
      if (this.peek().type === 'paren' && this.peek().value === '(') {
        this.consume()
        const args: AstNode[] = []
        if (!(this.peek().type === 'paren' && this.peek().value === ')')) {
          args.push(this.parseOr())
          while (this.peek().type === 'comma') {
            this.consume()
            args.push(this.parseOr())
          }
        }
        if (!(this.peek().type === 'paren' && this.peek().value === ')')) {
          throw new ExpressionError('Expected closing parenthesis')
        }
        this.consume()
        return { kind: 'call', name, args }
      }
      return { kind: 'identifier', name }
    }
    if (t.type === 'paren' && t.value === '(') {
      this.consume()
      const inner = this.parseOr()
      if (!(this.peek().type === 'paren' && this.peek().value === ')')) {
        throw new ExpressionError('Expected closing parenthesis')
      }
      this.consume()
      return inner
    }
    throw new ExpressionError(`Unexpected token '${t.value}'`)
  }
}

function resolveIdentifier(name: string, ctx: EvaluationContext): unknown {
  if (name in ctx.participant_responses) return ctx.participant_responses[name]
  if (name in ctx.panel_metadata) return ctx.panel_metadata[name]
  if (name in ctx.system_variables) return ctx.system_variables[name]
  return undefined
}

function toNumber(v: unknown): number {
  if (typeof v === 'number') return v
  if (typeof v === 'boolean') return v ? 1 : 0
  const n = Number(v)
  return Number.isFinite(n) ? n : 0
}

function toBool(v: unknown): boolean {
  if (typeof v === 'boolean') return v
  if (v === null || v === undefined || v === '') return false
  if (Array.isArray(v)) return v.length > 0
  if (typeof v === 'number') return v !== 0
  return true
}

function evalNode(node: AstNode, ctx: EvaluationContext): unknown {
  switch (node.kind) {
    case 'literal':
      return node.value
    case 'identifier':
      return resolveIdentifier(node.name, ctx)
    case 'unary': {
      const v = evalNode(node.arg, ctx)
      if (node.op === 'not') return !toBool(v)
      if (node.op === '-') return -toNumber(v)
      throw new ExpressionError(`Unknown unary operator ${node.op}`)
    }
    case 'binary': {
      const l = evalNode(node.left, ctx)
      const r = evalNode(node.right, ctx)
      switch (node.op) {
        case '==':
          return String(l) === String(r)
        case '!=':
          return String(l) !== String(r)
        case '>':
          return toNumber(l) > toNumber(r)
        case '<':
          return toNumber(l) < toNumber(r)
        case '>=':
          return toNumber(l) >= toNumber(r)
        case '<=':
          return toNumber(l) <= toNumber(r)
        case '&&':
          return toBool(l) && toBool(r)
        case '||':
          return toBool(l) || toBool(r)
        case '+':
          return toNumber(l) + toNumber(r)
        case '-':
          return toNumber(l) - toNumber(r)
        case '*':
          return toNumber(l) * toNumber(r)
        case '/':
          return toNumber(l) / toNumber(r)
        default:
          throw new ExpressionError(`Unknown operator ${node.op}`)
      }
    }
    case 'call': {
      const args = node.args.map((a) => evalNode(a, ctx))
      switch (node.name) {
        case 'if':
          return toBool(args[0]) ? args[1] : args[2]
        case 'sum':
          return args.reduce<number>((acc, v) => acc + toNumber(v), 0)
        case 'count':
          return args.filter((v) => toBool(v)).length
        case 'rand': {
          const min = toNumber(args[0])
          const max = toNumber(args[1])
          return Math.floor(Math.random() * (max - min + 1)) + min
        }
        case 'array_filter': {
          const arr = args[0]
          if (!Array.isArray(arr)) return []
          const pred = args[1]
          return arr.filter((item) => String(item) === String(pred))
        }
        default:
          throw new ExpressionError(`Unknown function ${node.name}`)
      }
    }
    default:
      throw new ExpressionError('Invalid AST node')
  }
}

/** Evaluate a standalone ExpressionScript expression. */
export function evaluateExpression(
  expression: string,
  context: EvaluationContext | Record<string, unknown>,
): unknown {
  const trimmed = expression.trim()
  if (!trimmed) return true
  const ctx: EvaluationContext =
    'participant_responses' in context
      ? (context as EvaluationContext)
      : {
          participant_responses: context as Record<string, unknown>,
          panel_metadata: {},
          system_variables: {},
        }
  const tokens = tokenize(trimmed)
  const ast = new Parser(tokens).parse()
  return evalNode(ast, ctx)
}

/** Replace `{expression}` placeholders in survey text (micro-tailoring). */
export function interpolateText(template: string, context: EvaluationContext): string {
  return template.replace(/\{([^{}]+)\}/g, (_match, expr: string) => {
    try {
      const result = evaluateExpression(expr.trim(), context)
      if (result === null || result === undefined) return ''
      return String(result)
    } catch {
      return `{${expr}}`
    }
  })
}

/** Return true when relevance expression evaluates truthy (empty = always show). */
export function isRelevant(expression: string | null | undefined, context: EvaluationContext): boolean {
  if (!expression?.trim()) return true
  try {
    return toBool(evaluateExpression(expression, context))
  } catch {
    return false
  }
}
